pkg/controllers/namespace.go

```diff
@@ -0,0 +1,256 @@
+		return
+	}
+
+	for i := range c.eventHandlers {
+		klog.V(4).Infof("Calling handler.OnNamespaceAdd")
+		c.eventHandlers[i].OnNamespaceAdd(ns)
+	}
+}
+
+// handleUpdateNamespace calls registered event handlers OnNamespaceUpdate
+func (c *NamespaceConfig) handleUpdateNamespace(oldObj, newObj interface{}) {
+	oldNamespace, ok := oldObj.(*v1.Namespace)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", oldObj))
+		return
+	}
+	ns, ok := newObj.(*v1.Namespace)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", newObj))
+		return
+	}
+	for i := range c.eventHandlers {
+		klog.V(4).Infof("Calling handler.OnNamespaceUpdate")
+		c.eventHandlers[i].OnNamespaceUpdate(oldNamespace, ns)
+	}
+}
+
+// handleDeleteNamespace calls registered event handlers OnNamespaceDelete
+func (c *NamespaceConfig) handleDeleteNamespace(obj interface{}) {
+	ns, ok := obj.(*v1.Namespace)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", obj))
+		}
+		if ns, ok = tombstone.Obj.(*v1.Namespace); !ok {
+			utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", obj))
+			return
+		}
+	}
+	for i := range c.eventHandlers {
+		klog.V(4).Infof("Calling handler.OnNamespaceDelete")
+		c.eventHandlers[i].OnNamespaceDelete(ns)
+	}
+}
+
+// NamespaceInfo contains information that defines a namespace.
+type NamespaceInfo struct {
+	Name   string
+	Labels map[string]string
+}
+
+// nsChange represents a change in Namespace represented via NamespaceMap
+type nsChange struct {
+	previous NamespaceMap
+	current  NamespaceMap
+}
+
+// NamespaceChangeTracker carries state about uncommitted changes to an arbitrary number of
+// Namespaces in the node, keyed by their namespace and name
+type NamespaceChangeTracker struct {
+	// lock protects items.
+	lock sync.Mutex
+	// items maps a service to its podChange.
+	items map[string]*nsChange
+}
+
+// newNamespaceInfo creates a new instance of NamespaceInfo
+func (nct *NamespaceChangeTracker) newNamespaceInfo(ns *v1.Namespace) *NamespaceInfo {
+	return &NamespaceInfo{
+		Name:   ns.Name,
+		Labels: ns.Labels,
+	}
+}
+
+// NewNamespaceChangeTracker creates a new instance of NamespaceChangeTracker
+func NewNamespaceChangeTracker() *NamespaceChangeTracker {
+	return &NamespaceChangeTracker{
+		items: make(map[string]*nsChange),
+	}
+}
+
+// nsToNamespaceMap creates NamespaceMap from v1.Namespace.
+// Note(adrianc): it is basically a map with single entry.
+func (nct *NamespaceChangeTracker) nsToNamespaceMap(ns *v1.Namespace) NamespaceMap {
+	if ns == nil {
+		return nil
+	}
+
+	namespaceMap := make(NamespaceMap)
+	nsInfo := nct.newNamespaceInfo(ns)
+	namespaceMap[ns.Name] = *nsInfo
+	return namespaceMap
+}
+
+// Update handles an update of a given Namespace
+func (nct *NamespaceChangeTracker) Update(previous, current *v1.Namespace) bool {
+	ns := current
+
+	if nct == nil {
```

Comment:
technically yes.

https://go.dev/play/p/ukxM_WdZWoy

Reply:
this can be removed as it should never be nil. (and program should panic imo)

the only benefit may be for UT since the Update call will be a noop. (which is the only method that the server is going to use).
that said, i dont think this should be relied on for UT.